### ISSUE TITLE
How to select a model checkpoint with best epoch loss after pretraining

### ISSUE THREAD

**Issue by blazejdolicki** [facebookresearch/vissl#554]
## ❓ How to do something using VISSL

Describe what you want to do, including:
1. what inputs you will provide, if any:
2. what outputs you are expecting:

When pretraining a model, I save a checkpoint after every epoch. Once the training is done, I would like to select the checkpoint with the best epoch loss to use it for linear evaluation. Currently, there are no logs telling me what is the loss at the end of the epoch. The closest to what I'm looking for is the `stdout.json` file that looks like this (first 3 lines):
```
{"Rank": 0, "ep": 0, "iter": 0, "lr": 0.03, "loss": 6.54846, "btime(ms)": 0, "eta": "0:00:00", "peak_mem(M)": 8184, "weight_decay": [0.0001, 0.0]}
{"Rank": 0, "ep": 0, "iter": 1, "lr": 0.03, "loss": 6.87759, "btime(ms)": 23220, "eta": "55 days, 0:57:14", "peak_mem(M)": 8184, "weight_decay": [0.0001, 0.0], "max_iterations": 204800}
{"Rank": 0, "ep": 0, "iter": 5, "lr": 0.03, "loss": 7.50373, "btime(ms)": 4831, "eta": "11 days, 10:51:21", "peak_mem(M)": 8184, "weight_decay": [0.0001, 0.0]}
```
However, when I look into `vissl/hooks/log_hooks.py` the line assigning the value of the loss above is 
```
loss_val = round(task.last_batch.loss.data.cpu().item(), 5)
```
So from the name `task.last_batch` I presume this loss is the loss just from the last iteration/batch and not from the whole epoch. To sum up, the output I would like is a file with loss after every epoch (calculated over all examples, not just from one batch) and/or a model checkpoint from the best phase wrt loss, something like `model_best_phase.torch`. Therefore, I would like to ask if there's some easy way to do it for example by changing the config, and if not, given your knowledge about the library, could you recommend which file in the source code should I modify to obtain the functionality I'm looking for?

For a simple PyTorch loop this is what I do
```
    epoch_losses = []
    for epoch_idx in tqdm(range(args.num_epochs)):  # loop over the dataset multiple times
            # train model for one epoch
            # save loss
            epoch_losses.append(epoch_loss)

            # set new best loss and epoch and save model if current loss is better than the best one so far
            if best_epoch_loss < current_epoch_loss:
                best_epoch_loss  = current_epoch_loss

                # save parameters of the best model for inference
                logging.info(f"Saving the best model to {best_model_path}")
                torch.save(model.state_dict(), best_model_path)
    # save epoch_losses to file
```
But how could I modify VISSL to achieve the same?

## ❓ What does an API do and how to use it?
Please link to which API or documentation you're asking about from
https://github.com/facebookresearch/vissl/tree/main/docs

Not sure what to answer here, maybe it's related to hooks https://github.com/facebookresearch/vissl/blob/main/docs/source/api/hooks.rst

NOTE:

1. Only general answers are provided.
   If you want to ask about "why X did not work", please use the
   [Unexpected behaviors](https://github.com/facebookresearch/vissl/issues/new/choose) issue template.

2. About how to implement new models / new dataloader / new training logic, etc., check documentation first.

4. We do not answer general machine learning / computer vision questions that are not specific to VISSL, such as how a model works, how to improve your training/make it converge, or what algorithm/methods can be used to achieve X.

**Comment by blazejdolicki** [comment]
Thanks for your response, I think for now I will select the epoch based on losses from stdout or possibly just pick the final checkpoint because in my experiments so far the loss constantly decreases.
